subprotocols/src/expression.rs

```diff
@@ -0,0 +1,166 @@
+use std::sync::Arc;
+
+use ff_ext::ExtensionField;
+
+mod evaluate;
+mod op;
+
+mod macros;
+
+pub type Point<E> = Arc<Vec<E>>;
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub enum Constant {
+    /// Base field
+    Base(i64),
+    /// Challenge
+    Challenge(usize),
```

Comment:
It's a bit weird to put challenge in constant, but ok in this stage.
For later, I would propose to extend [expression](https://github.com/scroll-tech/ceno/blob/master/ceno_zkvm/src/expression.rs#L668) in ceno_zkvm so they can be well support for pre-compile * ceno_zkvm. We did lots of optimisation on ceno_zkvm expression, e.g. we will optimize scalar & poly multiplication, e.g. `poly_1 * c1 * c2 * c3 ` into `poly * ((c1 * c2) * c3)`, for latter just O(N) with smaller constant factor

Reply:
Sure, I agree. There is also an `Expression` in `mpcs` and it's better to unify the three structs :p.